service/history/historyEngine.go

```diff
@@ -2563,7 +2562,7 @@ func (e *historyEngineImpl) failWorkflowTask(
 	}
 
 	if _, err = mutableState.AddWorkflowTaskFailedEvent(
-		scheduleID, startedID, cause, failure, request.GetIdentity(), request.GetBinaryChecksum(), "", "", 0,
+		scheduleID, startedID, workflowTaskFailedErr.failedCause, failure.NewServerFailure(workflowTaskFailedErr.Error(), true), request.GetIdentity(), request.GetBinaryChecksum(), "", "", 0,
```

Comment:
Why are we wrapping it in `NewServerFailure`?

Reply:
Maybe we can add another method on workflowTaskFailedErr type to convert it into right format to add an history event for it.

Reply:
`WorkflowTaskFailedEvent` has failure field which may contains different failures. Here it is just error message but other places may set other failures. The best way to refactor this, is to replace `workflowTaskFailedError` with `WorkflowTaskFailure` which will contains `failedCause` enum and error message. I don't know if it worth it. I would leave it as is for now.